### PR TITLE
feat(APP-2627): Add internationalisation support for ODS library

### PR DESCRIPTION
**Pull request by rcaragon**
## Description

This pull request organizes texts into a single source of truth and allows users of the library to customize copy for both `code` and `module` packages.

Task: [APP-2627](https://aragonassociation.atlassian.net/browse/APP-2627)

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

-   [x] I have selected the correct base branch.
-   [x] I have performed a self-review of my own code.
-   [x] I have commented my code, particularly in hard-to-understand areas.
-   [x] I have made corresponding changes to the documentation.
-   [x] My changes generate no new warnings.
-   [x] Any dependent changes have been merged and published in downstream modules.
-   [x] I ran all tests with success and extended them if necessary.
-   [x] I have updated the `CHANGELOG.md` file in the root folder of the package after the [UPCOMING] title and before
        the latest version.
-   [ ] I have tested my code on the test network.

[APP-2627]: https://aragonassociation.atlassian.net/browse/APP-2627?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/core/index.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+export * from './assets/copy';
```
Comment (cgero-eth) [inline src/core/index.ts:1]:
```suggestion
export * from './assets';
```

File: src/core/assets/copy/IOdsCoreCopy.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Interface representing the structure of copy texts used in various parts of the ODS Core package.
+ * Each property in the interface corresponds to a specific component or feature, containing the necessary
+ * text labels or functions that return text strings.
+ */
+
+export interface IOdsCoreCopy {
```
Comment (cgero-eth) [inline src/core/assets/copy/IOdsCoreCopy.ts:7]: What do you think if instead we remove this interface, rename `enCopy` to `coreCopy` and use `typeof coreCopy` on the core provider? In that way we won't need to maintain the interface when adding changes to the coreCopy

File: src/core/assets/copy/copy.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { type IOdsCoreCopy } from './IOdsCoreCopy';
```
Comment (cgero-eth) [inline src/core/assets/copy/copy.ts:1]: If we apply the suggestion above, I'd rename this file to `coreCopy.ts` (or `enCopy.ts` if we keep it named as `enCopy`)

File: src/core/components/odsCoreProvider/odsCoreProvider.mdx [hunk 4]
```diff
@@ -37,7 +38,3 @@ export const Application = () => {
     )
 }
 ```
-
-Refer to the component's properties below for more details on the default configuration objects used for each provider.
-
-<ArgTypes of={OdsCoreProvider} />
```
Comment (cgero-eth) [inline src/core/components/odsCoreProvider/odsCoreProvider.mdx:43]: Why has this block been removed?

File: src/core/components/odsCoreProvider/odsCoreProvider.mdx [hunk 5]
```diff
@@ -5,12 +5,13 @@ import { OdsCoreProvider } from './odsCoreProvider';
 
 <Title>OdsCoreProvider</Title>
 
-The `<OdsCoreProvider />` is a React context that defines the image and link elements to be rendered across all the ODS
-components, defaulting such values to `<img />` and `<a />` tags. The provider is used by the `<AvatarBase />` and
-`<LinkBase />` components, which are in turn used by all the other core and module components.
+The `<OdsCoreProvider />` is a React context that defines the image, link elements, and copy texts to be rendered across
+all the ODS components, defaulting such values to `<img />`, `<a />` tags, and a default set of copy texts. The provider
+is used by the `<AvatarBase />` and `<LinkBase />` components, which are in turn used by all the other core and module
```
Comment (cgero-eth) [inline src/core/components/odsCoreProvider/odsCoreProvider.mdx:10]: > The provider is used by the `<AvatarBase />` and `<LinkBase />` component

This is wrong now, the provider is used by many more components

File: src/modules/components/address/addressInput/addressInput.tsx [hunk 6]
```diff
@@ -18,6 +18,7 @@ import {
 import type { IWeb3ComponentProps } from '../../../types';
 import { addressUtils, ensUtils } from '../../../utils';
 import { MemberAvatar } from '../../member';
+import { useOdsModulesContext } from '../../odsModulesProvider/odsModulesProvider';
```
Comment (cgero-eth) [inline src/modules/components/address/addressInput/addressInput.tsx:21]: All `odsModulesProvider` imports can/should be changed to:
```suggestion
import { useOdsModulesContext } from '../../odsModulesProvider';
```

File: src/modules/assets/copy/copy.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    majorityVotingResult: {
+        winningOption: 'Winning Option',
+        stage: 'Stage',
+    },
+    proposalDataListItemStatus: {
+        youHaveVoted: "You've voted",
+    },
+    proposalDataListItemStructure: {
+        by: 'By',
+        creatorsNumber: (count) => `${count}+ ${count > 1 ? 'creators' : 'creator'}`,
```
Comment (cgero-eth) [inline src/modules/assets/copy/copy.ts:31]: We probably don't need to handle a singular case here, `creatorsNumber` should actually be `maxCreatorsNumber` and there's no case where `count` should be set to `0`. Even `0+` might still be more than one 😄

File: src/modules/components/proposal/proposalDataListItem/proposalDataListItemStructure/proposalDataListItemStructure.tsx [hunk 8]
```diff
@@ -71,8 +74,10 @@ export const ProposalDataListItemStructure: React.FC<IProposalDataListItemStruct
                         'min-h-5 gap-x-0.5 text-sm leading-tight text-neutral-600 md:min-h-6 md:gap-x-1 md:text-base',
                     )}
                 >
-                    By
-                    {showParsedPublisher === false && <span>3+ creators</span>}
+                    {copy.proposalDataListItemStructure.by}
+                    {showParsedPublisher === false && (
+                        <span>{copy.proposalDataListItemStructure.creatorsNumber(3)}</span>
```
Comment (cgero-eth) [inline src/modules/components/proposal/proposalDataListItem/proposalDataListItemStructure/proposalDataListItemStructure.tsx:79]:
```suggestion
                        <span>{copy.proposalDataListItemStructure.creatorsNumber(maxPublishersDisplayed)}</span>
```

File: src/modules/assets/copy/copy.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        stage: 'Stage',
+        of: 'of',
+        members: 'Members',
+    },
+    majorityVotingResult: {
+        winningOption: 'Winning Option',
+        stage: 'Stage',
+    },
+    proposalDataListItemStatus: {
+        youHaveVoted: "You've voted",
```
Comment (cgero-eth) [inline src/modules/assets/copy/copy.ts:27]: In that way, if we later change it to `you did vote` or `voted successfully`, the label still makes sense
```suggestion
        voted: "You've voted",
```

File: src/modules/components/proposal/proposalDataListItem/approvalThresholdResult/approvalThresholdResult.tsx [hunk 10]
```diff
@@ -27,9 +29,9 @@ export const ApprovalThresholdResult: React.FC<IApprovalThresholdResultProps> =
                 <span className="text-primary-400">
                     {formatterUtils.formatNumber(approvalAmount, { format: NumberFormat.GENERIC_SHORT })}
                 </span>
-                <span>of</span>
+                <span>{copy.approvalThresholdResult.of}</span>
```
Comment (cgero-eth) [inline src/modules/components/proposal/proposalDataListItem/approvalThresholdResult/approvalThresholdResult.tsx:32]: Could we also do a `outOf` copy here? Like:
```
const formattedApprovalThreshold = {formatterUtils.formatNumber(approvalThreshold, { format: NumberFormat.GENERIC_SHORT })};

[...]

<span>{copy.approvalThresholdResult.outOf(formattedApprovalThreshold)</span>
```

File: src/modules/components/odsModulesProvider/odsModulesProvider.tsx [hunk 11]
```diff
@@ -37,20 +45,51 @@ export interface IOdsModulesProviderProps {
...
+            <WagmiProvider config={wagmiConfig}>
+                <QueryClientProvider client={queryClient}>
+                    <OdsCoreProvider values={coreProviderValues}>{children}</OdsCoreProvider>
+                </QueryClientProvider>
+            </WagmiProvider>
+        </odsModulesContext.Provider>
     );
 };
+
+export const useOdsModulesContext = (): Required<IOdsModulesContext> => {
```
Comment (cgero-eth) [inline src/modules/components/odsModulesProvider/odsModulesProvider.tsx:91]: The `required` is probably not needed, right?
```suggestion
export const useOdsModulesContext = (): IOdsModulesContext => {
```

Comment (cgero-eth) [inline src/modules/components/odsModulesProvider/odsModulesProvider.tsx:91]: Could you add a simple test to the `useOdsModulesContext` hook like we already do for the `useOdsCoreContext`?

**Review by cgero-eth — COMMENTED** [review]
Thanks for the changes @rcaragon!

File: src/core/assets/copy/IOdsCoreCopy.ts (hunk 2 above)
Comment (rcaragon) [reply src/core/assets/copy/IOdsCoreCopy.ts]: Good idea 👍, updated.

File: src/core/assets/copy/copy.ts (hunk 3 above)
Comment (rcaragon) [reply src/core/assets/copy/copy.ts]: ✅ https://github.com/aragon/ods/pull/221/commits/e5e0ecb68c8c0f30b1055fb3c24e8e9b02be13a6

File: src/core/components/odsCoreProvider/odsCoreProvider.mdx (hunk 4 above)
Comment (rcaragon) [reply src/core/components/odsCoreProvider/odsCoreProvider.mdx]: It was an accident, fixed https://github.com/aragon/ods/pull/221/commits/213b022b7d7184c0d1624222ffd3e6e5e01530e0

File: src/core/components/odsCoreProvider/odsCoreProvider.mdx (hunk 5 above)
Comment (rcaragon) [reply src/core/components/odsCoreProvider/odsCoreProvider.mdx]: Removed this sentence all together https://github.com/aragon/ods/pull/221/commits/4889d2d8f80058bbc1615400862db79407c6e8f4

File: src/modules/components/address/addressInput/addressInput.tsx (hunk 6 above)
Comment (rcaragon) [reply src/modules/components/address/addressInput/addressInput.tsx]: ✅ https://github.com/aragon/ods/pull/221/commits/4e26e3dd159f2792c0c16d5ff696d111f9c25635

File: src/modules/assets/copy/copy.ts (hunk 7 above)
Comment (rcaragon) [reply src/modules/assets/copy/copy.ts]: ✅ https://github.com/aragon/ods/pull/221/commits/5442722929ba987b81593c7ae4ddbf400604f23a

File: src/modules/assets/copy/copy.ts (hunk 9 above)
Comment (rcaragon) [reply src/modules/assets/copy/copy.ts]: ✅ https://github.com/aragon/ods/pull/221/commits/e0774462dbc0f20790ca7353858bf1aaeee95943

File: src/modules/components/proposal/proposalDataListItem/approvalThresholdResult/approvalThresholdResult.tsx (hunk 10 above)
Comment (rcaragon) [reply src/modules/components/proposal/proposalDataListItem/approvalThresholdResult/approvalThresholdResult.tsx]: ✅ https://github.com/aragon/ods/pull/221/commits/4284e4d6f87e2510798e8184d8d7d6444dc4d1f1

File: src/modules/components/odsModulesProvider/odsModulesProvider.tsx (hunk 11 above)
Comment (rcaragon) [reply src/modules/components/odsModulesProvider/odsModulesProvider.tsx]: ✅ https://github.com/aragon/ods/pull/221/commits/ba38b76702d85d1fdc5404a83db9d159de254e74

Comment (rcaragon) [reply src/modules/components/odsModulesProvider/odsModulesProvider.tsx]: ✅ https://github.com/aragon/ods/pull/221/commits/304bbc92e9dd0cf6cd8c484d67998ce6d895edf6

File: src/core/index.ts (hunk 1 above)
Comment (cgero-eth) [reply src/core/index.ts:1]: We already have a index file under `/assets`

File: src/modules/components/odsModulesProvider/odsModulesProvider.mdx [hunk 12]
```diff
@@ -15,7 +15,11 @@ wrapped using the following providers:
 -   `<WagmiProvider />`: needed to know which chain and RPC endpoint to use when fetching blockchain data;
 -   `<QueryClientProvider />`: needed to cache the data fetched from the blockchain;
 
-The provider also renders the `<OdsCoreProvider />` and passes the values set in the `coreProviderValues` property.
+The provider also renders the `<OdsCoreProvider />` and passes the values set in the `coreProviderValues` property and
+the values set in the `values` property for module-specific configurations.
```
Comment (cgero-eth) [inline src/modules/components/odsModulesProvider/odsModulesProvider.mdx:19]: Could you improve the readability here? Sounds like the provider passes the `values` prop to the `OdsCoreProvider`. I would document the module-specific props in a new sentence

Comment (rcaragon) [reply src/modules/components/odsModulesProvider/odsModulesProvider.mdx]: ✅ https://github.com/aragon/ods/pull/221/commits/f666bcf9caf68ae81d6d59c9342bf52a57d03f2d

**Review by cgero-eth — APPROVED** [review]

**Pull request merged**